src/stream.js

```diff
@@ -1,412 +1,416 @@
-"use strict";
+'use strict';
 
 function Lazy(func) {
-    this.has_evaluated = false;
-    this.func = func;
-    this.value = null;
+  this.has_evaluated = false;
+  this.func = func;
+  this.value = null;
 }
 
 Lazy.prototype = {
-    eval: function() {
-        if ( this.has_evaluated ) {
-            return this.value;
-        }
-
-        this.value = this.func();
-        this.has_evaluated = true;
-        return this.value;
+  eval: function() {
+    if (this.has_evaluated) {
+      return this.value;
     }
+
+    this.value = this.func();
+    this.has_evaluated = true;
+    return this.value;
+  }
 };
 
 function Eager(func) {
-    this.func = func;
+  this.func = func;
 }
 
 Eager.prototype = {
-    eval: function() {
-        return this.func();
-    }
+  eval: function() {
+    return this.func();
+  }
+}
+
+function Stream(head, tailPromise, wrapper) {
+  if (typeof wrapper == 'undefined') {
+    wrapper = Lazy;
+  }
+
+  this.wrapper = wrapper;
+
+  if (typeof head != 'undefined') {
+    this.headValue = head;
+  }
+  if (typeof tailPromise == 'undefined') {
+    // Ignore the next function from coverage as it is never called.
+    // Instead we cover this.create() to ensure that tailPromise
+    // contains the value we want.
+    /* istanbul ignore next */
+    tailPromise = function() {
+      return this.create();
+    };
+  }
+
+  if (typeof tailPromise == 'function') {
+    tailPromise = new wrapper(tailPromise);
+  }
+
+  this.tailPromise = tailPromise;
 }
 
-function Stream( head, tailPromise, wrapper ) {
-    if ( typeof wrapper == 'undefined' ) {
-        wrapper = Lazy;
+// TODO: write some unit tests
+Stream.prototype = {
```

Comment:
I wanted to say that you should set the `constructor` property back to `Stream` because you're overwriting by setting the `prototype` to `Object` but I can't  find a reason why that should matter. `new Stream() instanceof Stream` still works...

Reply:
@dotnetCarpenter do you referring to something like this:
```
Stream.prototype = {
  constructor: Steam,
  // ... other methods
}
```
I did not

Reply:
Yes. But I don't see the point of adding it to the prototype object. I don't know the practical application of the `constructor` property in JS - I just know it is there. So for my sake, we do not have to add it. Maybe I'm ignorant but unless you know why we should put it in the prototype - let's just leave it as it is.